Documentation/guides/Extending LiveCode.md

```diff
@@ -810,6 +830,77 @@ are available in the LiveCode source repository. This is a more
 efficient means of inclusion as it allows the compiler to strip unused 
 symbols.
 
+##### LiveCode's Foundation Library and the FFI
+
+LiveCode's foundation library, called [libfoundation](https://github.com/livecode/livecode/tree/develop/libfoundation),
+contains a number of useful helper functions that can aid writing widgets and
+libraries in LiveCode Builder; indeed, many of the included core LiveCode Builder
+API modules (com.livecode.file, com.livecode.arithmetic, etc) make extensive
+use of the foundation library.
+
+The first step is to determine if the function you wish to use is available to use
+via the FFI. To do this it is necessary to look at the [foundation library header file](https://github.com/livecode/livecode/blob/develop/libfoundation/include/foundation.h)
+If the function is declared in the header file and the declaration is preceded
+by `MC_DLLEXPORT` then that function can be accessed through the FFI.
+
+For instance, the `MCMemoryNew` function has the following declaration in the header
+file:
+
+      MC_DLLEXPORT bool MCMemoryNew(size_t size, void*& r_record);
+and so it can be accessed through the API.
+
+###### Wrapping a Foundation Library Function in the FFI
+
+To provide an example, the `MCStringDecode` function is a function that converts a variable
+from the type `Data` to the native LiveCode string. It has the following declaration in the
+foundation library header file:
+
+      MC_DLLEXPORT bool MCStringDecode(MCDataRef data, MCStringEncoding encoding, bool is_external_rep, MCStringRef& r_string);
+
+To wrap the function in LiveCode Builder it is necessary to understand the parameters
+of the function:
+
+* the `data` parameter is the variable of type `Data` that is to be converted to a `String`
+* the `encoding` parameter is an integer between zero and eleven that specifies the type of
+  data to be decoded. Of interest for this example is that `4` represents UTF8 encoded data
+  (the other available formats can be found in the `MCStringEncoding` declaration in the 
+  foundation library header file
+* the `is_external_rep` parameter exists for legacy reasons only, and takes a value `false`
+* the `r_string` parameter is the resulting decoded string. Unlike all the other parameters
+  it should be declared as an `out` parameter because it gives the decoded String
+
+The method to bind to a foundation library function is the same as binding to any C function,
+with two important distinctions:
+
+* the name of the function in the LCB declaration must exactly match the name of the function
+  in the foundation library
+* the binding string is simply `<builtin>`, with no class or function name specified
```

Comment:
Is it worthwhile explaining that any of `<builtin>`, `MCStringDecode` and `c:MCStringDecode` will work as a binding string here? Perhaps it's worth saying that if you want to name your foreign handler something other than the name of the function then you can use a binding string.

Reply:
I'll add in a note about `c:MCStringDecode`. I don't think it is necessary to explain `MCStringDecode` because this is something that is not specific to binding to the foundation library and should instead be explained with the general format of C binding strings (probably in the language reference).